Use opmath_type for CUDA logcumsumexp

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83425

This improves precision by reducing the number of narrowing
conversions, as well as reducing compile times from 2m 30s to 1m 25s
on my machine.

peterbell10 committed:
Use accscalar_t for CUDA logcumsumexp

This improves precision by reducing the number of narrowing
conversions, as well as reducing compile times from 2m 30s to 1m 25s
on my machine.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/LogcumsumexpKernel.cu:
```diff
@@ -18,10 +18,11 @@ void launch_logcumsumexp_cuda_kernel(const TensorBase& result, const TensorBase&
       [&]() {
         using accscalar_t = acc_type<scalar_t, true>;
```
better to use opmath_t here

peterbell10 committed:
Update on "Use accscalar_t for CUDA logcumsumexp"


This improves precision by reducing the number of narrowing
conversions, as well as reducing compile times from 2m 30s to 1m 25s
on my machine.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "Use opmath_type for CUDA logcumsumexp"


This improves precision by reducing the number of narrowing
conversions, as well as reducing compile times from 2m 30s to 1m 25s
on my machine.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -l

ngimel commented:
@pytorchbot -g

ngimel commented:
@pytorchbot merge -g

huydhn commented:
@pytorchbot revert -m 'This break ROCm trunk test https://hud.pytorch.org/pytorch/pytorch/commit/06a64f7eaa47ce430a3fa61016010075b59b18a7' -c gha

huydhn commented:
@pytorchbot revert -m 'This break ROCm trunk test https://hud.pytorch.org/pytorch/pytorch/commit/06a64f7eaa47ce430a3fa61016010075b59b18a7' -c nosignal

huydhn commented:
@ngimel I'm trying to revert this PR because this seems to break rocm tests on trunk

```
======================================================================
FAIL [0.060s]: test_logcumsumexp_cuda (__main__.TestTorchDeviceTypeCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1921, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1921, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 988, in wrapper
    fn(*args, **kwargs)
  File "test_torch.py", line 2486, in test_logcumsumexp
    self.assertEqual(expected, actual)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2418, in assertEqual
    msg=(lambda generated_msg: f"{generated_msg} : {msg}") if isinstance(msg, str) and self.longMessage else msg,
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 2 / 9 (22.2%)
Greatest absolute difference: nan at index (7,) (up to 1e-05 allowed)
Greatest relative difference: nan at index (7,) (up to 1.3e-06 allowed)
```

peterbell10 committed:
Update on "Use opmath_type for CUDA logcumsumexp"


This improves precision by reducing the number of narrowing
conversions, as well as reducing compile times from 2m 30s to 1m 25s
on my machine.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -g
